lib/auth/auth_with_roles.go

```diff
@@ -3970,10 +3970,32 @@ func (a *ServerWithRoles) DeleteNamespace(name string) error {
 
 // GetRoles returns a list of roles
 func (a *ServerWithRoles) GetRoles(ctx context.Context) ([]types.Role, error) {
-	if err := a.action(apidefaults.Namespace, types.KindRole, types.VerbList, types.VerbRead); err != nil {
+	authErr := a.action(apidefaults.Namespace, types.KindRole, types.VerbList, types.VerbRead)
+	if authErr == nil {
+		return a.authServer.GetRoles(ctx)
+	}
+
+	roles, err := a.authServer.GetRoles(ctx)
+	if err != nil {
+		// If we get an error here, let's return the previous auth error.
+		return nil, trace.Wrap(authErr)
+	}
+
+	var filteredRoles []types.Role
+	// See if the user has access to these roles through things like where clauses.
+	for _, role := range roles {
+		sctx := &services.Context{User: a.context.User, Resource: role}
+		if err := a.actionWithContext(sctx, apidefaults.Namespace, types.KindRole, types.VerbList, types.VerbRead); err == nil {
```

Comment:
It's not super clear from this naming but does `actionWithContext` respect "where" rules and `action` doesn't, or what's the difference between them?

Reply:
It's not so much using `actionWithContext` as it is supplying the `sctx`, which contains a `services.Context` with both the user and the `Resource` supplied. With the `Resource` supplied, objects can be evaluated in where clauses.